### PR TITLE
Fix issue #1045 (XmlReader.ReadAsync goes into infinite recursion)

### PR DESCRIPTION
**Pull request by krwq**
If the file seems to have no changes, the actual change is that I changed encoding to utf-8 without BOM.

When we removed Uri support we removed a piece of code which was handling URIs in XmlTextReaderImpl. This code should've stayed as XmlResolver should decide if URIs are supported or not.

### CODE REVIEW

File: src/System.Xml.ReaderWriter/tests/XmlReader/packages.config [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="System.Runtime" version="4.0.20-beta-22703" />
+  <package id="System.Runtime.Extensions" version="4.0.10-beta-22703" />
```
Comment (krwq) [inline src/System.Xml.ReaderWriter/tests/XmlReader/packages.config]: undo this

**Comment by krwq** [comment]
@dotnet-bot test this please

File: src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            }
+        }
+
+        [Fact]
+        public static void ReadAsyncAfterInitializationWithUriThrows()
+        {
+            using (XmlReader reader = XmlReader.Create("http://test.test/test.html", new XmlReaderSettings() { Async = true }))
+            {
+                Exception e = Assert.Throws<AggregateException>(() => reader.ReadAsync().Wait());
+                Assert.IsType<XmlException>(e.InnerException);
```
Comment (stephentoub) [inline src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs]: These two lines could just be:
```C#
Assert.Throws<XmlException>(() => reader.ReadAsync().GetAwaiter().GetResult());
```

File: src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        <d>ddd</d>
+                    </c>
+                </root>";
+
+        private static Stream GetDummyXmlStream()
+        {
+            byte[] buffer = Encoding.UTF8.GetBytes(_dummyXml);
+            MemoryStream ms = new MemoryStream(buffer);
+            if (ms.Position != 0)
+                throw new NotImplementedException();
```
Comment (stephentoub) [inline src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs]: In what circumstance would this happen?

File: src/System.Xml.ReaderWriter/src/System/Xml/Core/XmlTextReaderImplAsync.cs [hunk 4]
```diff
@@ -74,6 +76,43 @@ private Task FinishInitAsync()
             }
         }
 
+        private async Task FinishInitUriStringAsync()
+        {
+            // Substitutes removed XmlResolver.GetEntityAsync
+            Task<object> getEntityAsync = Task.Run(() =>
```
Comment (stephentoub) [inline src/System.Xml.ReaderWriter/src/System/Xml/Core/XmlTextReaderImplAsync.cs]: I realize you're just restoring existing code, but do you know why that code was using Task.Run here?  It's scheduling a task that it just then awaits.

File: src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs (hunk 3 above)
Comment (krwq) [inline src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs]: That is leftover from development where I wasn't sure if Position is 0 and just threw some random exception. Thanks for finding it!

File: src/System.Xml.ReaderWriter/src/System/Xml/Core/XmlTextReaderImplAsync.cs (hunk 4 above)
Comment (krwq) [inline src/System.Xml.ReaderWriter/src/System/Xml/Core/XmlTextReaderImplAsync.cs]: Will change, thanks! I'm actually not sure if the tasks are run in separate threads. For Thread this is obvious, task is confusing to me.

File: src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs (hunk 2 above)
Comment (krwq) [inline src/System.Xml.ReaderWriter/tests/XmlReader/Tests/AsyncReaderLateInitTests.cs]: Will change

**Comment by krwq** [comment]
I addressed the feedback.

**Comment by krwq** [comment]
@dotnet-bot, test this please

**Comment by ellismg** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**